src/runtime/refiner.ts

```diff
@@ -8,141 +8,709 @@
         }
-        return this.fn(exprs);
+        break;
+      }
+    }
+    this.segments.splice(i, j-i, new Segment(x, y));
+  }
+
+  intersectWithSeg(seg: Segment): void {
+    const newRange = new Range();
+    for (const subRange of this.segments) {
+      if (subRange.overlapsWith(seg)) {
+        newRange.segments.push(Segment.overlap(seg, subRange));
+      }
+    }
+    this._segments = newRange.segments;
+  }
+
+  static makeInitialGivenOp(op: string, val: ExpressionPrimitives): Range {
+    switch (op) {
+      case Op.LT: return new Range([Segment.openOpen(Number.NEGATIVE_INFINITY, val)]);
+      case Op.LTE: return new Range([Segment.openClosed(Number.NEGATIVE_INFINITY, val)]);
+      case Op.GT: return new Range([Segment.openOpen(val, Number.POSITIVE_INFINITY)]);
+      case Op.GTE: return new Range([Segment.closedOpen(val, Number.POSITIVE_INFINITY)]);
+      case Op.EQ: return new Range([Segment.closedClosed(val, val)]);
+      case Op.NEQ: return Range.complementOf(new Range([Segment.closedClosed(val, val)]));
+      default: throw new Error(`Unsupported operator: field ${op} number`);
+    }
+  }
+
+  static updateGivenOp(op: string, ranges: Range[]): Range {
+    switch (op) {
+      case Op.AND: {
+        return Range.intersectionOf(ranges[0], ranges[1]);
+      }
+      case Op.OR: {
+        return Range.unionOf(ranges[0], ranges[1]);
+      }
+      case Op.EQ: {
+        const lc = Range.complementOf(ranges[0]);
+        const rc = Range.complementOf(ranges[1]);
+        const lnr = Range.intersectionOf(ranges[0], ranges[1]);
+        const lcnrc = Range.intersectionOf(lc, rc);
+        return Range.unionOf(lnr, lcnrc);
+      }
+      case Op.NEQ: {
+        const lc = Range.complementOf(ranges[0]);
+        const rc = Range.complementOf(ranges[1]);
+        const lnrc = Range.intersectionOf(ranges[0], rc);
+        const lcnr = Range.intersectionOf(lc, ranges[1]);
+        return Range.unionOf(lnrc, lcnr);
+      }
+      case Op.NOT: {
+        return Range.complementOf(ranges[0]);
+      }
+      default: throw new Error(`Unsupported operator: cannot update range`);
     }
+  }
+}
+
+export class Segment {
+  from: Boundary;
+  to: Boundary;
+
+  constructor(from: Boundary, to: Boundary) {
+    if (!Segment.isValid(from, to)) {
+      throw new Error(`Invalid range from: ${from}, to:${to}`);
+    }
+    this.from = {...from};
+    this.to = {...to};
+  }
+
+  static isValid(from: Boundary, to: Boundary): boolean {
+    if (to.val < from.val) {
+      return false;
+    } else if (from.val === to.val && (from.kind === 'open' || to.kind === 'open')) {
+      return false;
+    }
+    return true;
+  }
+
+  static closedClosed(from: number, to: number): Segment {
+    const seg = new Segment({val: from, kind: 'closed'}, {val: to, kind: 'closed'});
+    return seg;
+  }
+
+  static openOpen(from: number, to: number): Segment {
+    const seg = new Segment({val: from, kind: 'open'}, {val: to, kind: 'open'});
```

Comment:
Nit: (for clean up later, not now), just return these, no need to name a variable and then return it.